frame/support/src/storage/mod.rs

```diff
@@ -321,10 +321,18 @@ pub trait IterableStorageMap<K: FullEncode, V: FullCodec>: StorageMap<K, V> {
 	/// this, you'll get undefined results.
 	fn iter() -> Self::Iterator;
 
+	/// Enumerate all elements in the map starting from a specified `starting_key` in no particular
+	/// order. If you alter the map while doing this, you'll get undefined results.
+	fn iter_from(starting_key: Vec<u8>) -> Self::Iterator;
```

Comment:
The doc should precise that it iterates after the starting_key, the iterator will never return the (key/value) pair corresponding to the starting_key.

same for other methods

Also I think we should be more precise on the doc for iterator, this doesn't iterate in no particular order, it iterates on the lexicographic order of the encoded key. This what allows us to resume iterator

Also maybe it is better to name it `starting_raw_key` to convey the sense that this is the final key: encoded key with storage prefixes.

Reply:
Is it 100% accurate that it iterates on the lexicographical order? Because I tried calling `next()` on an iterator in the test, and the element that gets returned seemed to be random. Concretely speaking, I had a `MyMap` that has numbered keys from 1 to 10, calling `next()` on the iterator yielded the `6` key, which does not look to be lexicographically ordered at all.

Reply:
it should be the lexicographic order of the encoded key.
So for the keys represented by array of byte: `[1, 255, 3, 1]`, `[1, 255]`, `[1, 255, 4, 0]`, the order must be
 `[1, 255]`, `[1, 255, 3, 1]`, `[1, 255, 4, 0]`.
 
the doc for next key say so https://substrate.dev/rustdocs/latest/sp_io/storage/fn.next_key.html

Reply:
> Concretely speaking, I had a MyMap that has numbered keys from 1 to 10, calling next() on the iterator yielded the 6 key, which does not look to be lexicographically ordered at all.

It is because the hashes don't follow the same order as the key.

Reply:
> it should be the lexicographic order of the encoded and hashed key. (the final key put in the trie)
So for the keys represented by array of byte: [1, 255, 3, 1], [1, 255], [1, 255, 4, 0], the order must be
[1, 255], [1, 255, 3, 1], [1, 255, 4, 0].

^ not sure what convention is, but imo would be nice to include that in the docs too and a note pointing out

> the hashes don't follow the same order as the key.

For me at least, this bits of info connected the missing pieces to understanding the iteration order